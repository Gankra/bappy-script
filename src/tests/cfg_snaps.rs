fn dump_cfg(input: &str) -> String {
    let mut program = crate::Program::typed(input);
    program.parse().unwrap();
    program.check();

    let ctx = program.ctx.unwrap();
    let cfg = program.cfg.unwrap();

    cfg.format(&ctx).unwrap()
}

#[test]
fn capture_basics() {
    let program = r#"
    struct Point {
        x: Int,
        y: Int,
    }
    
    let pt = Point { x: 2, y: 7 }
    fn captures(arg: Int) -> Int {
        ret add(arg, pt.y)
    }

    fn no_captures(arg: Int) -> Int {
        ret add(arg, arg)
    }

    fn calls_funcs() -> Int {
        let a = captures(23)
        let b = no_captures(29)
        ret add(a, b)
    }

    let my_fn = captures
    let my_fn2 = no_captures

    print captures(2)
    print my_fn(3)

    print no_captures(6)
    print my_fn2(7)

    print calls_funcs()

    let x = 1
    let y = true
    let z = (x, y)
    if z.1 {
        print add(1, add(x, 2))
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("basics", cfg);
}

#[test]
fn loop_basics() {
    let program = r#"
    let cond = true
    let mut counter = 3
    loop {
        if eq(counter, 0) {
            break
        }
        set counter = sub(counter, 1)
        print counter

        loop {
            print "hello"
            if cond {
                break
            } else {
                continue
            }
            print "unreachable"
        }
        print "looped!"
    }
    ret 0
"#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("loop_basics", cfg);
}

#[test]
fn func_names_are_paths() {
    let program = r#"
    fn do_a_compy(x: Int) -> Int {
        ret mul(x, x)
    }
    let z = 5
    fn captured_do_a_compy(x: Int) -> Int {
        ret do_a_compy(add(x, z))
    }
    struct MyClosure {
        func: fn(Int) -> Int
        capture: Int
    }
    fn call_closure(close: MyClosure) -> Int {
        ret close.func(close.capture)
    }
    
    let close1 = MyClosure { func: do_a_compy, capture: 7 }
    let close2 = MyClosure { func: captured_do_a_compy, capture: 9 }
    
    print call_closure(close1)
    print call_closure(close2)
    
    ret close2.capture
    "#;

    let cfg = dump_cfg(program);
    insta::assert_snapshot!("func_names_are_paths", cfg);
}
