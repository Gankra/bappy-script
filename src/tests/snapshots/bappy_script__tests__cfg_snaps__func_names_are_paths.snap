---
source: src/tests/cfg_snaps.rs
assertion_line: 117
expression: cfg

---
#struct0_MyClosure {
  func: fn (Int) -> Int
  capture: Int
}

#struct1_MyClosure {
  func: fn (Int) -> Int
  capture: Int
}

[intrinsic] #fn0_add(lhs, rhs)
[intrinsic] #fn1_sub(lhs, rhs)
[intrinsic] #fn2_mul(lhs, rhs)
[intrinsic] #fn3_eq(lhs, rhs)
[intrinsic] #fn4_not(rhs)

#fn5_main():
  bb0():
    %1 = Int(5)
    %2 = (%1,)
    %3 = heap_alloc()
    %3.* = %2
    %4 = [closure](#fn7_captured_do_a_compy, %3)
    %5 = #fn6_do_a_compy
    %6 = Int(7)
    %7 = #struct0_MyClosure{ 0: %5, 1: %6 }
    %8 = %4
    %9 = Int(9)
    %10 = #struct1_MyClosure{ 0: %8, 1: %9 }
    %11 = %7
    %12 = #fn8_call_closure(%11)
    print %12
    %13 = %10
    %14 = #fn8_call_closure(%13)
    print %14
    %15 = %10.1
    *ret = %15
    %16 = Int(1)
    jmp bb1(%16)

  bb1(%0):
    drop %10
    drop %7
    drop %4
    ret


#fn6_do_a_compy(%0_x: Int):
  bb0(%0):
    %2 = #fn2_mul(%0, %0)
    *ret = %2
    %3 = Int(1)
    jmp bb1(%3)

  bb1(%1):
    ret


#fn7_captured_do_a_compy(%0_x: Int, %6_[closure]: &(Int,)):
  bb0(%0, %6):
    %2 = %6.*.0
    %3 = #fn0_add(%0, %2)
    %4 = #fn6_do_a_compy(%3)
    *ret = %4
    %5 = Int(1)
    jmp bb1(%5)

  bb1(%1):
    ret


#fn8_call_closure(%0_close: MyClosure):
  bb0(%0):
    %2 = %0.1
    %3 = (%0.0)(%2)
    *ret = %3
    %4 = Int(1)
    jmp bb1(%4)

  bb1(%1):
    ret


